zed: detect and offline physically removed devices

loli10K opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Close #1537 

### Description
<!--- Describe your changes in detail -->
~~This commit adds a new test case to the ZFS Test Suite to verify ZED can detect when a cache device is physically removed from a running system.~~ ~~The underlying issue has been (probably) fixed for quite some time: this change, if anything, can only help prevent regression in functionality.~~

EDIT: The existing functionality is not reliable on some kernels (https://github.com/zfsonlinux/zfs/pull/7926#issuecomment-423657451), additional code has been added to implement this in userland (using events coming from the existing libudev monitor thread). ZED will now offline devices (data and l2arc, but not spares[1]) when they cannot be handled by a spare in the pool.

NOTE 1: we cannot offline a spare device.

NOTE 2: a modified "TEST" file will be pushed initially to "stress test" this change and avoid flaky scripts in the ZTS.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
Testing done in https://github.com/zfsonlinux/zfs/issues/1537#issuecomment-422262928
Test added to the ZFS Test Suite.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).
- [ ] Change has been approved by a ZFS on Linux member.

loli10K commented:
This seems to work only on some kernels: testing was initially done on 4.6.4 where removing a disk (`echo 1 > /sys/block/<dev>/device/delete`) results in the following code being called:

```
l2arc_feed_thread() -> ZIO pipeline -> vdev_disk_io_done()
```

https://github.com/zfsonlinux/zfs/blob/b8a90418f3a9c23b89c5d2c729a4dd0fea644508/module/zfs/vdev_disk.c#L804-L812

then since `zio->io_error == EIO` is true and `check_disk_change()` returns `1` we call:

```
spa_async_request(SPA_ASYNC_REMOVE) -> ... -> vdev_set_state(VDEV_STATE_REMOVED)
```

We don't even need zed running to detect the disk as "REMOVED" as all the work is done in-kernel:

```
root@linux:~# stap -v -d kernel -d spl -d zfs -e '
probe
module("zfs").function("zfs_post_common").call
{
   printf(" -> %s spa_name=%s name=%s\n", ppfunc(), $spa->spa_name$, $name$$);
   print_backtrace();
}
probe
module("zfs").function("vdev_set_state").call
{
   printf(" <- %s vdev_guid=%s, state=%s\n", ppfunc(), $vd->vdev_guid$$, $state$$);
   print_backtrace();
}'
Pass 1: parsed user script and 113 library scripts using 94348virt/38824res/5108shr/34264data kb, in 330usr/0sys/327real ms.
Pass 2: analyzed script: 2 probes, 11 functions, 0 embeds, 0 globals using 99680virt/45816res/6620shr/39596data kb, in 290usr/300sys/602real ms.
Pass 3: using cached /root/.systemtap/cache/ca/stap_cad4333257455369b06b33ef3278df06_7146.c
Pass 4: using cached /root/.systemtap/cache/ca/stap_cad4333257455369b06b33ef3278df06_7146.ko
Pass 5: starting run.
 <- vdev_set_state vdev_guid=16043073036057741552, state=3
 0xffffffffc06bde10 : vdev_set_state+0x0/0x2a0 [zfs]
 0xffffffffc06947a3 : spa_async_remove+0x73/0xb0 [zfs]
 0xffffffffc06a440b : spa_async_thread+0x33b/0x430 [zfs]
 0xffffffffc02facae : thread_generic_wrapper+0x7e/0xc0 [spl]
 0xffffffff8109ac4f : kthread+0xdf/0x100 [kernel]
 0xffffffff815cc3f2 : ret_from_fork+0x22/0x40 [kernel]
 0xffffffff8109ab70 : kthread+0x0/0x100 [kernel] (inexact)
 -> zfs_post_common spa_name="testpool" name="statechange"
 0xffffffffc06f9570 : zfs_post_common.constprop.2+0x0/0x30 [zfs]
 0xffffffffc06f962b : zfs_post_state_change+0x8b/0xc0 [zfs]
 0xffffffffc06bdf01 : vdev_set_state+0xf1/0x2a0 [zfs]
 0xffffffffc06947a3 : spa_async_remove+0x73/0xb0 [zfs]
 0xffffffffc06a440b : spa_async_thread+0x33b/0x430 [zfs]
 0xffffffffc02facae : thread_generic_wrapper+0x7e/0xc0 [spl]
 0xffffffff8109ac4f : kthread+0xdf/0x100 [kernel]
 0xffffffff815cc3f2 : ret_from_fork+0x22/0x40 [kernel]
 0xffffffff8109ab70 : kthread+0x0/0x100 [kernel] (inexact)
^CPass 5: run completed in 40usr/140sys/23568real ms.
root@linux:~# zpool status -g
  pool: testpool
 state: ONLINE
status: One or more devices has been removed by the administrator.
	Sufficient replicas exist for the pool to continue functioning in a
	degraded state.
action: Online the device using 'zpool online' or replace the device with
	'zpool replace'.
  scan: none requested
config:

	NAME                    STATE     READ WRITE CKSUM
	testpool                ONLINE       0     0     0
	  10508245030006472519  ONLINE       0     0     0
	cache
	  16043073036057741552  REMOVED      0     0     0
	spares
	  270949342587202867    AVAIL   

errors: No known data errors
root@linux:~# pgrep zed
root@linux:~# 
```

However on some other kernels (tested on 3.10.0-693.21.1.el7.x86_64) `check_disk_change()` seems to return `0` for removed disks:

```
[root@centos zfs]# stap -v -DMAXSTRINGLEN=1024 -d kernel -d spl -d zfs -e '
> probe
> kernel.function("check_disk_change").call
> {
>    printf(" -> %s %s\n", ppfunc(), $bdev->bd_disk$$);
>    print_backtrace();
> }
> probe
> kernel.function("check_disk_change").return
> {
>    printf(" <- %s %s\n", ppfunc(), $return$$);
> }'
Pass 1: parsed user script and 470 library scripts using 228276virt/41328res/3332shr/38064data kb, in 440usr/30sys/465real ms.
Pass 2: analyzed script: 2 probes, 38 functions, 0 embeds, 0 globals using 268840virt/82948res/4372shr/78628data kb, in 800usr/140sys/941real ms.
Pass 3: translated to C into "/tmp/stapmcexAG/stap_2a0f1db85741810451944afa5b77d2c4_10066_src.c" using 268840virt/83296res/4720shr/78628data kb, in 2860usr/150sys/3009real ms.
Pass 4: compiled C into "stap_2a0f1db85741810451944afa5b77d2c4_10066.ko" in 7710usr/690sys/7528real ms.
Pass 5: starting run.
[ ... ]
 -> check_disk_change {.major=8, .first_minor=16, .minors=16, .disk_name="sdb", .devnode=0x0, .events=0, .async_events=0, .part_tbl=0xffff88008ca82900, .part0={.start_sect=0, .nr_sects=0, .nr_sects_seq={.sequence=0}, .alignment_offset=0, .discard_alignment=0, .__dev={.parent=0xffff8800b495f948, .p=0xffff8800b4868b40, .kobj={.name="sdb", .entry={.next=0xffff88003644e088, .prev=0xffff88003644e088}, .parent=0x0, .kset=0xffff88013b117c60, .ktype=0xffffffff81ad4920, .sd=0x0, .kref={.refcount={.counter=4}}, .state_initialized=1, .state_in_sysfs=0, .state_add_uevent_sent=1, .state_remove_uevent_sent=1, .uevent_suppress=0}, .init_name=          (null), .type=0xffffffff81ab8f80, .mutex={.count={.counter=1}, .wait_lock={<union>={.rlock={.raw_lock={.val={.counter=0}}}}}, ...}, ...}, ...}, ...}
 0xffffffff81242c60 : check_disk_change+0x0/0x70 [kernel]
 0xffff880137a8a300
 0xffffffffc0bf65e9 : zio_vdev_io_done+0x129/0x250 [zfs] (inexact)
 0xffffffffc0bf8014 : __zio_execute+0x484/0x4d0 [zfs] (inexact)
 0xffffffff8102954d : __switch_to+0xcd/0x500 [kernel] (inexact)
 0xffffffffc0bf8107 : zio_execute+0xa7/0x1d0 [zfs] (inexact)
 0xffffffff810c0bd4 : __wake_up+0x44/0x50 [kernel] (inexact)
 0xffffffffc044ef2b : taskq_thread+0x2eb/0x610 [spl] (inexact)
 0xffffffff810c7c70 : default_wake_function+0x0/0x20 [kernel] (inexact)
 0xffffffffc0bf8060 : zio_execute+0x0/0x1d0 [zfs] (inexact)
 0xffffffffc044ec40 : taskq_thread+0x0/0x610 [spl] (inexact)
 0xffffffff810b4031 : kthread+0xd1/0xe0 [kernel] (inexact)
 0xffffffff810b3f60 : kthread+0x0/0xe0 [kernel] (inexact)
 0xffffffff816c0577 : ret_from_fork+0x77/0xb0 [kernel] (inexact)
 0xffffffff810b3f60 : kthread+0x0/0xe0 [kernel] (inexact)
 <- check_disk_change 0
```

We could theoretically respond to the "resource.fs.zfs.removed" event (which i believe comes from libudev) and try to offline the device instead of discarding it:

```
[root@centos zfs]# gdb -q zed -ex 'b zfs_retire_recv' -ex 'run -Ffv'
Reading symbols from /usr/sbin/zed...done.
Breakpoint 1 at 0x40cee0: file agents/zfs_retire.c, line 285.
Starting program: /usr/sbin/zed -Ffv
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Registered zedlet "scrub_finish-notify.sh"
Registered zedlet "resilver_finish-start-scrub.sh"
Registered zedlet "resilver_finish-notify.sh"
Registered zedlet "statechange-notify.sh"
Registered zedlet "all-syslog.sh"
Ignoring "zed.rc": not executable by user
Registered zedlet "statechange-led.sh"
Registered zedlet "vdev_clear-led.sh"
Registered zedlet "vdev_attach-led.sh"
Registered zedlet "pool_import-led.sh"
Ignoring "zed-functions.sh": not executable by user
Registered zedlet "data-notify.sh"
ZFS Event Daemon 0.8.0-rc1_9_g145c88f (PID 26426)
[ ... ]
zed_disk_event:
	class: EC_dev_remove
	subclass: disk
	dev_name: /dev/sdb1
	path: /devices/pci0000:00/0000:00:09.0/ata4/host3/target3:0:0/3:0:0:0/block/sdb/sdb1
	devid: ata-QEMU_HARDDISK_2-part1
	phys_path: pci-0000:00:09.0-ata-2.0
agent post event: mapping 'EC_dev_remove' to 'resource.fs.zfs.removed'
Diagnosis Engine: discarding 'resource.fs.zfs.removed for vdev 0
```

This needs more debugging.

behlendorf commented:
@loli10K extending the `zed` to handle this case would be pretty reasonable.  I would think it could also let us cleanly handle the case where non-l2arc devices are removed from an otherwise idle pool.

loli10K commented:
Yes, but for non-l2arc/non-spare devices we should probably replace them with a spare instead of offlining the device. Right now i'm facing a couple of issues with this change:

 * AFAIK spares cannot be offlined/onlined, so if zed receives a `resource.fs.zfs.removed` for a spare i don't know what the best course of action would be. We could probably try to `zpool_vdev_fault()`/`zpool_vdev_degrade()` the spare instead of `zpool_vdev_offline()` but then we would need to be able to "clean up" this if/when the spare is physically attached again.

 * some builders (amzn2, debian9) fails to pass the new test case due to a race condition: `zed`, somehow, is receiving the `EC_DEV_REMOVE` event from udev before the pool is even created and the SPA in present in the namespace:

http://build.zfsonlinux.org/builders/Amazon%202%20x86_64%20Release%20%28TEST%29/builds/2669/steps/shell_9/logs/log:

```
9:57:47.55 NOTE: Performing test-fail callback (/usr/share/zfs/zfs-tests/callbacks/zfs_dbgmsg.ksh)
19:57:47.55 =================================================================
19:57:47.55  Tailing last 200 lines of zfs_dbgmsg log
19:57:47.55 =================================================================
19:57:47.55 8 0 0x01 -1 0 857464022085 922919608137
19:57:47.55 timestamp    message 
19:57:47.55 1537819004   zfs_ioctl.c:1706:zfs_ioc_pool_configs(): [14979] zfs_ioc_pool_configs -> 
19:57:47.55 1537819004   zfs_ioctl.c:1716:zfs_ioc_pool_configs(): [14979] zfs_ioc_pool_configs <- (0)
19:57:47.55 1537819004   zfs_ioctl.c:1706:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs -> 
19:57:47.55 1537819004   zfs_ioctl.c:1708:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs <- (EEXIST)
19:57:47.55 1537819005   zfs_ioctl.c:1706:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs -> 
19:57:47.55 1537819005   zfs_ioctl.c:1708:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs <- (EEXIST)
19:57:47.55 1537819005   zfs_ioctl.c:1706:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs -> 
19:57:47.55 1537819005   zfs_ioctl.c:1708:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs <- (EEXIST)
19:57:47.55 1537819005   zfs_ioctl.c:1706:zfs_ioc_pool_configs(): [14981] zfs_ioc_pool_configs -> 
19:57:47.55 1537819005   zfs_ioctl.c:1716:zfs_ioc_pool_configs(): [14981] zfs_ioc_pool_configs <- (0)
19:57:47.55 1537819005   zfs_ioctl.c:1706:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs -> 
19:57:47.55 1537819005   zfs_ioctl.c:1716:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs <- (0)
19:57:47.55 1537819005   zfs_ioctl.c:1706:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs -> 
19:57:47.55 1537819005   zfs_ioctl.c:1716:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs <- (0)
19:57:47.55 1537819006   zfs_ioctl.c:1706:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs -> 
19:57:47.55 1537819006   zfs_ioctl.c:1708:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs <- (EEXIST)
19:57:47.55 1537819006   zfs_ioctl.c:1563:zfs_ioc_pool_create(): [16271] zfs_ioc_pool_create -> 
19:57:47.55 1537819006   vdev_disk.c:212:vdev_elevator_switch(): Unable to set "noop" scheduler for /dev/sda1 (sda): 256
19:57:47.55 1537819006   vdev_disk.c:212:vdev_elevator_switch(): Unable to set "noop" scheduler for /dev/sda1 (sda): 256
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 create pool version 5000; software version 5000/5; uts ip-172-31-7-133.us-west-1.compute.internal 4.14.67-71.56.amzn2.x86_64 #1 SMP Tue Sep 4 21:37:05 UTC 2018 x86_64
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@async_destroy=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@empty_bpobj=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@lz4_compress=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@multi_vdev_crash_dump=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@spacemap_histogram=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@enabled_txg=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@hole_birth=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@extensible_dataset=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@embedded_data=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@bookmarks=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@filesystem_limits=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@large_blocks=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@large_dnode=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@sha512=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@skein=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@edonr=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@userobj_accounting=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@encryption=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@project_quota=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@device_removal=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@obsolete_counts=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@spacemap_v2=enabled
19:57:47.55 1537819006   spa_history.c:318:spa_history_log_sync(): txg 4 set feature@allocation_classes=enabled
19:57:47.55 1537819006   zfs_ioctl.c:1636:zfs_ioc_pool_create(): [16271] zfs_ioc_pool_create <- (0) 
19:57:47.55 1537819006   spa_history.c:305:spa_history_log_sync(): command: zpool create testpool /mnt/file-vdev cache sda
19:57:47.55 1537819006   zfs_ioctl.c:1706:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs -> 
19:57:47.55 1537819006   zfs_ioctl.c:1716:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs <- (0)
19:57:47.55 1537819007   vdev_disk.c:144:vdev_disk_error(): zio error=5 type=2 offset=4194304 size=512 flags=808c0
19:57:47.55 1537819007   vdev_disk.c:808:vdev_disk_io_done(): check_disk_change(sda) = 0
19:57:47.55 1537819007   spa.c:7110:spa_async_request(): spa=testpool async request task=4
19:57:47.55 1537819007   vdev_disk.c:144:vdev_disk_error(): zio error=5 type=1 offset=270336 size=8192 flags=b08c1
19:57:47.55 1537819007   vdev_disk.c:144:vdev_disk_error(): zio error=5 type=1 offset=257433600 size=8192 flags=b08c1
19:57:47.55 1537819007   vdev_disk.c:144:vdev_disk_error(): zio error=5 type=1 offset=257695744 size=8192 flags=b08c1
19:57:47.55 1537819007   vdev_disk.c:808:vdev_disk_io_done(): check_disk_change(sda) = 0
19:57:47.55 1537819007   vdev_disk.c:808:vdev_disk_io_done(): check_disk_change(sda) = 0
19:57:47.55 1537819007   vdev_disk.c:808:vdev_disk_io_done(): check_disk_change(sda) = 0
19:57:47.55 1537819007   vdev.c:115:vdev_dbgmsg(): disk vdev '/dev/sda1': failed probe
19:57:47.55 1537819007   zfs_ioctl.c:1706:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs -> 
19:57:47.55 1537819007   zfs_ioctl.c:1716:zfs_ioc_pool_configs(): [14980] zfs_ioc_pool_configs <- (0)
19:57:47.55 1537819067   zfs_ioctl.c:1706:zfs_ioc_pool_configs(): [18361] zfs_ioc_pool_configs -> 
19:57:47.55 1537819067   zfs_ioctl.c:1716:zfs_ioc_pool_configs(): [18361] zfs_ioc_pool_configs <- (0)
19:57:47.55 1537819067   zfs_ioctl.c:1706:zfs_ioc_pool_configs(): [18362] zfs_ioc_pool_configs -> 
19:57:47.55 1537819067   zfs_ioctl.c:1716:zfs_ioc_pool_configs(): [18362] zfs_ioc_pool_configs <- (0)
19:57:47.55 1537819067   zfs_ioctl.c:1706:zfs_ioc_pool_configs(): [18363] zfs_ioc_pool_configs -> 
19:57:47.55 1537819067   zfs_ioctl.c:1716:zfs_ioc_pool_configs(): [18363] zfs_ioc_pool_configs <- (0)
```

We have multiple `libzfs_config:namespace_reload()` calls before `zfs_ioc_pool_create()`: they are from `zed` which is trying to get the vdev_tree to map "devid"->"guid" (which fails because we have no pools in the namespace):

```
19:57:47.54 --- being raw udev ---
19:57:47.54 nvlist version: 0
19:57:47.54 	devid = scsi-33333333000001770-part1
19:57:47.54 	dev_name = /dev/sda1
19:57:47.54 	path = /devices/pseudo_0/adapter0/host2/target2:0:0/2:0:0:0/block/sda/sda1
19:57:47.54 	is_slice = 1
19:57:47.54 	dev_size = 0xf600000
19:57:47.54 ---  end raw udev  ---
19:57:47.54 agent post event: mapping 'EC_dev_remove' to 'resource.fs.zfs.removed'
19:57:47.54 --- being payload ---
19:57:47.54 nvlist version: 0
19:57:47.54 	devid = scsi-33333333000001770-part1
19:57:47.54 	dev_name = /dev/sda1
19:57:47.54 	path = /devices/pseudo_0/adapter0/host2/target2:0:0/2:0:0:0/block/sda/sda1
19:57:47.54 	is_slice = 1
19:57:47.54 	dev_size = 0xf600000
19:57:47.54 	class = resource.fs.zfs.removed
19:57:47.54 	pool_guid = 0x0 <---------------------- ### no pool/vdev_tree information ###
19:57:47.54 	vdev_guid = 0x0 <---------------------- ### no pool/vdev_tree information ###
19:57:47.54 	time = 1537819005 749381
19:57:47.54 ---  end payload  ---
```

zed log:

```
19:57:47.54 Ignoring "vdev_id.conf": not executable by user
19:57:47.54 Ignoring "zed.rc": not executable by user
19:57:47.54 Ignoring "zed-functions.sh": not executable by user
19:57:47.54 Registered zedlet "resilver_finish-start-scrub.sh"
19:57:47.54 Ignoring "zed.debug.log": not executable by user
19:57:47.54 Ignoring "zed.log": not executable by user
19:57:47.54 ZFS Event Daemon 0.8.0-rc1_13_g04031d8f0 (PID 14973)
19:57:47.54 Add Agent: init
19:57:47.54 Diagnosis Engine: register module
19:57:47.54 Retire Agent: register module
19:57:47.54 zed_disk_event_init
19:57:47.54 Processing events since eid=0
19:57:47.54 namespace_reload ->
19:57:47.54 [14973] zpool_iter: 0 pools
19:57:47.54 Waiting for new udev disk events...
19:57:47.54 zed_udev_monitor: /dev/loop0 no devid source
19:57:47.54 zed_udev_monitor: /dev/loop2 no devid source
19:57:47.54 zed_udev_monitor: skip /dev/xvdb (in use by ext3)
19:57:47.54 zed_udev_monitor: /dev/loop1 no devid source
19:57:47.54 zed_disk_event:
19:57:47.54 	class: EC_dev_add
19:57:47.54 	subclass: disk
19:57:47.54 	dev_name: /dev/sda
19:57:47.54 	path: /devices/pseudo_0/adapter0/host2/target2:0:0/2:0:0:0/block/sda
19:57:47.54 	devid: scsi-33333333000001770
19:57:47.54 	dev_size: 268435456
19:57:47.54 zfs_slm_event: EC_dev_add.disk
19:57:47.54 zfs_deliver_add: adding scsi-33333333000001770 (NULL) (is_slice 0)
19:57:47.54 namespace_reload ->
19:57:47.54 namespace_reload: namespace hasn't changed
19:57:47.54 [14973] zpool_iter: 0 pools
19:57:47.54 zed_udev_monitor: skip /dev/xvda1 (in use by xfs)
19:57:47.54 zed_disk_event:
19:57:47.54 	class: EC_dev_status
19:57:47.54 	subclass: dev_dle
19:57:47.54 	dev_name: /dev/sda
19:57:47.54 	path: /devices/pseudo_0/adapter0/host2/target2:0:0/2:0:0:0/block/sda
19:57:47.54 	devid: scsi-33333333000001770
19:57:47.54 	dev_size: 268435456
19:57:47.54 zfs_slm_event: EC_dev_status.dev_dle
19:57:47.54 zfs_deliver_dle: no guid or physpath
19:57:47.54 namespace_reload ->
19:57:47.54 namespace_reload: namespace hasn't changed
19:57:47.54 [14973] zpool_iter: 0 pools
19:57:47.54 zfs_deliver_dle: device '' not found
19:57:47.54 zed_disk_event:
19:57:47.54 	class: EC_dev_add
19:57:47.54 	subclass: disk
19:57:47.54 	dev_name: /dev/sda1
19:57:47.54 	path: /devices/pseudo_0/adapter0/host2/target2:0:0/2:0:0:0/block/sda/sda1
19:57:47.54 	devid: scsi-33333333000001770-part1
19:57:47.54 	dev_size: 257949696
19:57:47.54 zfs_slm_event: EC_dev_add.disk
19:57:47.54 zfs_deliver_add: adding scsi-33333333000001770-part1 (NULL) (is_slice 1)
19:57:47.54 namespace_reload ->
19:57:47.54 namespace_reload: namespace hasn't changed
19:57:47.54 [14973] zpool_iter: 0 pools
19:57:47.54 zed_disk_event:
19:57:47.54 	class: EC_dev_remove
19:57:47.54 	subclass: disk
19:57:47.54 	dev_name: /dev/sda1
19:57:47.54 	path: /devices/pseudo_0/adapter0/host2/target2:0:0/2:0:0:0/block/sda/sda1
19:57:47.54 	devid: scsi-33333333000001770-part1
19:57:47.54 	dev_size: 257949696
19:57:47.54 zfs_agent_post_event: search by devid=scsi-33333333000001770-part1
19:57:47.54 namespace_reload ->
19:57:47.54 [14973] zpool_iter: 0 pools
19:57:47.54 zpool_iter: 0
19:57:47.54 --- being raw udev ---
19:57:47.54 nvlist version: 0
19:57:47.54 	devid = scsi-33333333000001770-part1
19:57:47.54 	dev_name = /dev/sda1
19:57:47.54 	path = /devices/pseudo_0/adapter0/host2/target2:0:0/2:0:0:0/block/sda/sda1
19:57:47.54 	is_slice = 1
19:57:47.54 	dev_size = 0xf600000
19:57:47.54 ---  end raw udev  ---
19:57:47.54 agent post event: mapping 'EC_dev_remove' to 'resource.fs.zfs.removed'
19:57:47.54 --- being payload ---
19:57:47.54 nvlist version: 0
19:57:47.54 	devid = scsi-33333333000001770-part1
19:57:47.54 	dev_name = /dev/sda1
19:57:47.54 	path = /devices/pseudo_0/adapter0/host2/target2:0:0/2:0:0:0/block/sda/sda1
19:57:47.54 	is_slice = 1
19:57:47.54 	dev_size = 0xf600000
19:57:47.54 	class = resource.fs.zfs.removed
19:57:47.54 	pool_guid = 0x0
19:57:47.54 	vdev_guid = 0x0
19:57:47.54 	time = 1537819005 749381
19:57:47.54 ---  end payload  ---
19:57:47.54 namespace_reload ->
19:57:47.54 [14973] zpool_iter: 0 pools
19:57:47.54 Diagnosis Engine: discarding 'resource.fs.zfs.removed for vdev 0
19:57:47.54 Retire Agent: zfs_retire_recv: 'resource.fs.zfs.removed'
19:57:47.54 namespace_reload ->
19:57:47.54 [14973] zpool_iter: 0 pools
19:57:47.54 zed_disk_event:
19:57:47.54 	class: EC_dev_add
19:57:47.54 	subclass: disk
19:57:47.54 	dev_name: /dev/sda1
19:57:47.54 	path: /devices/pseudo_0/adapter0/host2/target2:0:0/2:0:0:0/block/sda/sda1
19:57:47.54 	devid: scsi-33333333000001770-part1
19:57:47.54 	dev_size: 257949696
19:57:47.54 zfs_slm_event: EC_dev_add.disk
19:57:47.54 zfs_deliver_add: adding scsi-33333333000001770-part1 (NULL) (is_slice 1)
19:57:47.54 namespace_reload ->
19:57:47.54 namespace_reload: namespace hasn't changed
19:57:47.54 [14973] zpool_iter: 0 pools
19:57:47.54 zed_disk_event:
19:57:47.54 	class: EC_dev_status
19:57:47.54 	subclass: dev_dle
19:57:47.54 	dev_name: /dev/sda1
19:57:47.54 	path: /devices/pseudo_0/adapter0/host2/target2:0:0/2:0:0:0/block/sda/sda1
19:57:47.54 	devid: scsi-33333333000001770-part1
19:57:47.54 zfs_slm_event: EC_dev_status.dev_dle
19:57:47.54 zfs_deliver_dle: no guid or physpath
19:57:47.54 namespace_reload ->
19:57:47.54 [14973] zpool_iter: 1 pools
19:57:47.54 zfsdle_vdev_online: searching for '' in 'testpool'
19:57:47.54 zfs_deliver_dle: device '' not found
19:57:47.54 zed_udev_monitor: /dev/sda1 no devid source
19:57:47.54 zed_udev_monitor: /dev/sda1 no devid source
19:57:47.54 zed_udev_monitor: /dev/loop1 no devid source
19:57:47.54 zed_udev_monitor: skip /dev/xvdb (in use by ext3)
19:57:47.54 zed_udev_monitor: /dev/loop0 no devid source
19:57:47.54 zed_udev_monitor: /dev/loop2 no devid source
19:57:47.54 zed_udev_monitor: skip /dev/xvda1 (in use by xfs)
19:57:47.54 namespace_reload ->
19:57:47.54 [14973] zpool_iter: 1 pools
19:57:47.54 Diagnosis Engine: case opened (d4b13d01-ee0a-4588-9c42-df6689b69402)
19:57:47.54 Diagnosis Engine: opening case for vdev 6009880749961340520 due to 'ereport.fs.zfs.io'
19:57:47.54 Diagnosis Engine: error event 'ereport.fs.zfs.io'
19:57:47.54 Diagnosis Engine: error event 'ereport.fs.zfs.io'
19:57:47.54 Diagnosis Engine: error event 'ereport.fs.zfs.io'
19:57:47.54 Diagnosis Engine: error event 'ereport.fs.zfs.probe_failure'
```



Same issue in http://build.zfsonlinux.org/builders/Debian%209%20x86_64%20%28TEST%29/builds/1291/steps/shell_9/logs/log

loli10K commented on cmd/zed/agents/zfs_agents.c:
```diff
@@ -57,9 +57,11 @@ typedef struct guid_search {
 	uint64_t	gs_vdev_guid;
 	char		*gs_devid;
+	boolean_t	is_spare;
+	boolean_t	is_l2arc;
```
Note: these could probably be a single variable keeping track of the device type.

loli10K commented on cmd/zed/agents/zfs_retire.c:
```diff
@@ -318,6 +329,32 @@ zfs_retire_recv(fmd_hdl_t *hdl, fmd_event_t *ep, nvlist_t *nvl,
+
+			dev_name = zpool_vdev_name(NULL, zhp, vdev, B_FALSE);
+			fmd_hdl_debug(hdl, "zpool_vdev_offline '%s'", dev_name);
+			zpool_vdev_offline(zhp, dev_name, B_TRUE);
```
Note: this doesn't currently work (offlining a spare). Once resolved we could probably close https://github.com/zfsonlinux/zfs/issues/1105.

behlendorf commented on cmd/zed/agents/zfs_agents.c:
```diff
@@ -73,8 +80,29 @@ zfs_agent_iter_vdev(zpool_handle_t *zhp, nvlist_t *nvl, void *arg)
 	if (nvlist_lookup_nvlist_array(nvl, ZPOOL_CONFIG_CHILDREN,
 	    &child, &children) == 0) {
 		for (c = 0; c < children; c++)
```
nit: please add `{ }` for this loop

behlendorf commented on cmd/zed/agents/zfs_agents.c:
```diff
@@ -73,8 +80,29 @@ zfs_agent_iter_vdev(zpool_handle_t *zhp, nvlist_t *nvl, void *arg)
+	 */
+	if (nvlist_lookup_nvlist_array(nvl, ZPOOL_CONFIG_SPARES,
+	    &child, &children) == 0) {
+		for (c = 0; c < children; c++)
```
nit: please add `{ }` for this loop

behlendorf commented on cmd/zed/agents/zfs_agents.c:
```diff
@@ -73,8 +80,29 @@ zfs_agent_iter_vdev(zpool_handle_t *zhp, nvlist_t *nvl, void *arg)
+	}
+	if (nvlist_lookup_nvlist_array(nvl, ZPOOL_CONFIG_L2CACHE,
+	    &child, &children) == 0) {
+		for (c = 0; c < children; c++)
```
nit: please add `{ }` for this loop

behlendorf commented on cmd/zed/agents/zfs_agents.c:
```diff
@@ -53,13 +53,20 @@ pthread_t g_agents_tid;
 } guid_search_t;
 
-static void
+static boolean_t
```
It would be nice to have a block comment above this function describing its intended behavior.

behlendorf commented on cmd/zed/agents/zfs_agents.c:
```diff
@@ -53,13 +53,20 @@ pthread_t g_agents_tid;
+typedef enum device_type {
+	DEVICE_TYPE_L2ARC,
+	DEVICE_TYPE_SPARE,
+	DEVICE_TYPE_DISK
```
What do you think about renaming `DEVICE_TYPE_DISK` to `DEVICE_TYPE_PRIMARY` (or similar).  What's being determined here is if the vdev is part of the primary pool storage.

loli10K replied:
I couldn't think of a better name, i used `DEVICE_TYPE_DISK` (and `DEVICE_TYPE_L2ARC`/`DEVICE_TYPE_SPARE`) just because i found similarly named #defines here:

https://github.com/zfsonlinux/zfs/blob/54eb2c410ec8ef9f75c62e3d0e8f53ac8bdb5eea/include/sys/fs/zfs.h#L748-L754

`DEVICE_TYPE_PRIMARY` makes sense, i am going to use it.

behlendorf commented on cmd/zed/agents/zfs_agents.c:
```diff
@@ -169,12 +202,25 @@ zfs_agent_post_event(const char *class, const char *subclass, nvlist_t *nvl)
 		    FM_EREPORT_PAYLOAD_ZFS_POOL_GUID, pool_guid);
 		(void) nvlist_add_uint64(payload,
 		    FM_EREPORT_PAYLOAD_ZFS_VDEV_GUID, vdev_guid);
+		switch (devtype) {
```
It looks like your switch statement is missing `break;`'s.

loli10K commented on tests/zfs-tests/tests/functional/fault/auto_offline_001_pos.ksh:
```diff
@@ -0,0 +1,196 @@
+do
+	# 1. Create a pool with a spare
+	log_must zpool create -f $TESTPOOL $conf
+	# XXX: remove this, debugging centos-6 builder failure which cannot be reproduced on my environment
```
I think this is a race condition, the test seems to work on my CentOS6 builder at home; i regrettably spent more time debugging race conditions in the test case than writing code for this PR.

I am going to try a different approach and use `file_wait $ZED_LOG` instead of `zed_events_drain` since the latter does not wait for events coming from udev.

behlendorf commented on cmd/zed/agents/zfs_retire.c:
```diff
@@ -318,8 +337,21 @@ zfs_retire_recv(fmd_hdl_t *hdl, fmd_event_t *ep, nvlist_t *nvl,
+			fmd_hdl_debug(hdl, "zpool_vdev_offline '%s'", devname);
+			zpool_vdev_offline(zhp, devname, B_TRUE);
+		} else if (!fmd_prop_get_int32(hdl, "spare_on_remove") ||
+		    replace_with_spare(hdl, zhp, vdev) == B_FALSE) {
```
One alternative way to handle this would be to include a vdev's last expansion time in the config/vdev_tree.  Then the ZED could be updated to ignore "remove" event which immediately follow an expansion.  That's still not iron clad but it would cover the common case.

loli10K replied:
Pushed some prototype code to ignore "remove" events for expanding vdevs.

My favourite Debian home builder cannot run "zpool_expand_001_pos" so this was tested very briefly on a "spare" CentOS7 builder:

```
/usr/share/zfs/test-runner/bin/test-runner.py  -c /var/tmp/zfs-tests.498.run -T functional -i /usr/share/zfs/zfs-tests -I 1
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_expand/setup (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_expand/zpool_expand_001_pos (run as root) [00:13] [FAIL]
[   94.158494] BUG: unable to handle kernel paging request at ffffffffa0929000
[   94.159577] IP: [<ffffffffa0000fb4>] scsi_device_put+0x14/0x60 [scsi_mod]
[   94.160009] PGD 1816067 PUD 1817063 PMD d432b067 PTE 0
[   94.160009] Oops: 0000 [#1] SMP 
[   94.160009] Modules linked in: sd_mod crc_t10dif crct10dif_generic loop zfs(PO) icp(PO) zlua(PO) zcommon(PO) zunicode(PO) znvpair(PO) zavl(PO) spl(O) nfsd auth_rpcgss oid_registry nfs_acl nfs lockd fscache sunrpc crc32_pclmul aesni_intel aes_x86_64 lrw gf128mul glue_helper ablk_helper cryptd ppdev psmouse pcspkr serio_raw virtio_console i2c_piix4 evdev i2c_core processor parport_pc joydev parport pvpanic thermal_sys button virtio_balloon autofs4 ext4 crc16 mbcache jbd2 hid_generic usbhid hid sg sr_mod cdrom ata_generic virtio_scsi virtio_blk virtio_net crct10dif_pclmul crct10dif_common crc32c_intel ahci libahci floppy ata_piix xhci_hcd usbcore usb_common virtio_pci virtio_ring virtio libata scsi_mod [last unloaded: scsi_debug]
[   94.160009] CPU: 0 PID: 3544 Comm: systemd-udevd Tainted: P           O  3.16.0-4-amd64 #1 Debian 3.16.51-3
[   94.160009] Hardware name: Bochs Bochs, BIOS Bochs 01/01/2011
[   94.160009] task: ffff8800986ed450 ti: ffff8800d92dc000 task.ti: ffff8800d92dc000
[   94.160009] RIP: 0010:[<ffffffffa0000fb4>]  [<ffffffffa0000fb4>] scsi_device_put+0x14/0x60 [scsi_mod]
[   94.160009] RSP: 0018:ffff8800d92dfe68  EFLAGS: 00010292
[   94.160009] RAX: ffffffffa0929000 RBX: ffff880098480800 RCX: 0000000000000129
[   94.160009] RDX: 0000000000000000 RSI: 000000000002005d RDI: ffff880098473800
[   94.160009] RBP: ffff880098473800 R08: ffffffff81624a60 R09: 0000000000000000
[   94.160009] R10: ffff8800d9f3ed10 R11: 0000000000000246 R12: ffff880098480400
[   94.160009] R13: ffff8800d44d0ad8 R14: 000000000002005d R15: 0000000000000000
[   94.160009] FS:  00007f26f05fd880(0000) GS:ffff88011fc00000(0000) knlGS:0000000000000000
[   94.160009] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[   94.160009] CR2: ffffffffa0929000 CR3: 0000000098023000 CR4: 00000000000006f0
[   94.160009] Stack:
[   94.160009]  ffff880098480800 ffff880098473800 ffffffffa09337cb ffff8800d44d0ac0
[   94.160009]  0000000000000010 ffffffff811e4f7d ffff8800d9f3ed00 0000000000000010
[   94.160009]  ffff8800d8f9cd58 ffff88011a662960 ffff8800d44bb018 ffff8800d8f9cd58
[   94.160009] Call Trace:
[   94.160009]  [<ffffffffa09337cb>] ? scsi_disk_put+0x2b/0x40 [sd_mod]
[   94.160009]  [<ffffffff811e4f7d>] ? __blkdev_put+0x15d/0x1a0
[   94.160009]  [<ffffffff811e59d1>] ? blkdev_close+0x21/0x30
[   94.160009]  [<ffffffff811b0dda>] ? __fput+0xca/0x1d0
[   94.160009]  [<ffffffff8108838c>] ? task_work_run+0x8c/0xb0
[   94.160009]  [<ffffffff81014ef9>] ? do_notify_resume+0x69/0xa0
[   94.160009]  [<ffffffff81524fca>] ? int_signal+0x12/0x17
[   94.160009] Code: 5b 5d 41 5c 41 5d 41 5e 41 5f c3 66 66 66 2e 0f 1f 84 00 00 00 00 00 e8 fb 61 52 e1 55 48 89 fd 53 48 8b 07 48 8b 80 c0 00 00 00 <48> 8b 18 48 85 db 74 0d 48 89 df e8 dc a0 0d e1 48 85 c0 75 17 
[   94.160009] RIP  [<ffffffffa0000fb4>] scsi_device_put+0x14/0x60 [scsi_mod]
[   94.160009]  RSP <ffff8800d92dfe68>
[   94.160009] CR2: ffffffffa0929000
[   94.160009] ---[ end trace ed75d0bd63d1621a ]---
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_expand/cleanup (run as root) [00:00] [PASS]
```

loli10K commented:
> One question, was 3 seconds long enough in your testing. We may want to go longer here, or perhaps even scale this with the pool size, for very large pools we've definitely seen cases where a huge number of events can be generated and take a long time to process.

I think we might be safe here: the code deciding to discard "remove" events for recently expanded devices is executed in the "zed_udev_monitor" thread which should only process libudev events and doesn't have to deal with FM events. Slow/synchronous/blocking operations like faulting a device or activating hot spares run in another "zfs_agent_consumer_thread" thread and it doesn't seem they could delay us.

Unfortunately i don't own the hardware required to test this change on large pools with many devices.

I'm still investigating the CentOS 6 issue on a local builder.

loli10K commented on tests/zfs-tests/tests/functional/fault/auto_offline_001_pos.ksh:
```diff
@@ -0,0 +1,177 @@
+
+	# cleanup
+	destroy_pool $TESTPOOL
+	log_must parted "/dev/${removedev}" -s -- mklabel msdos
```
CentOS6 failure was caused by yet another race condition in udev events processing: the second iteration (first `for` loop) was getting "delete" events for partitions created on the first iteration. I added a `parted / mklabel msdos` and `block_device_wait` at the end of each loop.

Unfortunately we now have the Coverage builder failing "zpool_expand_001_pos": i am going to try and reproduce this on a local Ubuntu17 builder.

loli10K replied:
I wasn't able to reproduce "zpool_expand_001_pos" failure on a Ubuntu17 builder running this changes/PR, but i am seeing a different failure on a Debian9 builder running 6644e5bb6:

```
Test: /usr/share/zfs/zfs-tests/tests/functional/cli_root/zpool_expand/zpool_expand_001_pos (run as root) [00:28] [FAIL]
02:11:46.99 ASSERTION: zpool can be autoexpanded after set autoexpand=on on vdev expansion
02:11:46.99 NOTE: Setting up loopback, scsi_debug, and file vdevs
02:11:46.99 SUCCESS: truncate -s 1073741824 /var/tmp/vdev_lo
02:11:47.02 SUCCESS: losetup /dev/loop3 /var/tmp/vdev_lo
02:11:47.31 SUCCESS: modprobe scsi_debug dev_size_mb=1024 add_host=1 num_tgts=1 max_luns=1 sector_size=512 physblk_exp=0
02:11:48.31 SUCCESS: truncate -s 1073741824 /var/tmp/vdev_raw
02:11:51.02 SUCCESS: zpool create -f -o autoexpand=on testpool1 /dev/loop3 sda /var/tmp/vdev_raw
02:11:51.05 NOTE: Expanding loopback, scsi_debug, and file vdevs
02:11:51.06 SUCCESS: truncate -s 2147483648 /var/tmp/vdev_lo
02:11:51.06 SUCCESS: losetup -c /dev/loop3
02:11:57.66 SUCCESS: truncate -s 2147483648 /var/tmp/vdev_raw
02:12:12.50 SUCCESS: zpool online -e testpool1 /var/tmp/vdev_raw
02:12:12.81 NOTE: testpool1   has previous size: 3019898880 and  expanded size: 6241124352
02:12:13.14 pool testpool1 has not expanded,  1/3 vdevs expanded
02:12:13.14 NOTE: Performing test-fail callback (/usr/share/zfs/zfs-tests/callbacks/zfs_dbgmsg.ksh)
02:12:13.14 =================================================================
02:12:13.14  Tailing last 200 lines of zfs_dbgmsg log
02:12:13.14 =================================================================
[ ... cut previous runs ... ]
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 create pool version 5000; software version zfs-0.8.0-rc1-87-g6644e5bb6; uts linux 4.9.0-3-amd64 #1 SMP Debian 4.9.30-2+deb9u5 (2017-09-19) x86_64
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set autoexpand=1
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@async_destroy=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@empty_bpobj=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@lz4_compress=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@multi_vdev_crash_dump=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@spacemap_histogram=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@enabled_txg=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@hole_birth=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@extensible_dataset=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@embedded_data=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@bookmarks=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@filesystem_limits=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@large_blocks=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@large_dnode=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@sha512=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@skein=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@edonr=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@userobj_accounting=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@encryption=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@project_quota=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@device_removal=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@obsolete_counts=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@zpool_checkpoint=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@spacemap_v2=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@allocation_classes=enabled
02:12:13.15 1541319109   spa_history.c:319:spa_history_log_sync(): txg 4 set feature@resilver_defer=enabled
02:12:13.15 1541319111   spa_history.c:306:spa_history_log_sync(): command: zpool create -f -o autoexpand=on testpool1 /dev/loop3 sda /var/tmp/vdev_raw
02:12:13.15 1541319112   spa_history.c:306:spa_history_log_sync(): command: zfs set mountpoint=/var/tmp/testdir testpool/testfs
02:12:13.15 1541319114   spa.c:7112:spa_async_request(): spa=testpool1 async request task=1
02:12:13.15 1541319129   spa.c:7112:spa_async_request(): spa=testpool1 async request task=1
02:12:13.15 1541319129   spa_history.c:319:spa_history_log_sync(): txg 13 vdev online pool 'testpool1' size: 4093640704(+1073741824)
02:12:13.15 1541319129   spa_history.c:339:spa_history_log_sync(): ioctl reopen
02:12:13.15 1541319129   spa.c:7112:spa_async_request(): spa=testpool1 async request task=1
02:12:13.15 1541319132   spa.c:7112:spa_async_request(): spa=testpool1 async request task=1
02:12:13.15 1541319132   spa_history.c:319:spa_history_log_sync(): txg 18 vdev online pool 'testpool1' size: 6241124352(+2147483648)
02:12:13.15 1541319132   spa_history.c:306:spa_history_log_sync(): command: zpool online -e testpool1 /var/tmp/vdev_raw
02:12:13.15 1541319132   spa_history.c:339:spa_history_log_sync(): ioctl reopen
02:12:13.15 =================================================================
02:12:13.15  End of zfs_dbgmsg log
02:12:13.15 =================================================================
```

`testpool1` (stripe vdev) has successfully expanded from 3019898880 to 6241124352, but the test is only counting 1 expanded device:

> 02:12:13.14 pool testpool1 has not expanded,  **1**/3 vdevs expanded

when it should be counting **2** according to the dbgmsg output (the test greps `zpool history -il`):

> 02:12:13.15 1541319129   spa_history.c:319:spa_history_log_sync(): txg 13 vdev online pool 'testpool1' size: 4093640704(+1073741824)
> 02:12:13.15 1541319132   spa_history.c:319:spa_history_log_sync(): txg 18 vdev online pool 'testpool1' size: 6241124352(+2147483648)

I am going to bump the expansion timeout from 3 to 10 seconds and see what happens to the Coverage builder.

behlendorf replied:
That does appear to have helped.  Let's go with the longer time.

behlendorf commented:
@loli10K would you mind rebasing this one last time to resolve the trivial merge conflict.  Then I believe this is good to go.

don-brady commented on cmd/zed/agents/zfs_agents.c:
```diff
@@ -156,30 +199,55 @@ zfs_agent_post_event(const char *class, const char *subclass, nvlist_t *nvl)
 		(void) nvlist_add_uint64(payload,
 		    FM_EREPORT_PAYLOAD_ZFS_VDEV_GUID, vdev_guid);
-
-		(void) gettimeofday(&tv, NULL);
```
Was wondering why the FM_EREPORT_TIME was dropped from the record?  Is it recorded else where?

don-brady replied:
ah, then it's good to go.

behlendorf commented:
@loli10K no problem, thanks for doing the hard part!
